File: R/ai_insights.R
# AI-Powered Insights Module
# Supports Ollama (local) or OpenAI API

#' Generate AI insights from health data
#' Uses OpenAI API if OPENAI_API_KEY is set, otherwise tries Ollama
#' @param data_summary Text summary of the health data
#' @param user_context Optional context (age, country, etc.)
generate_ai_insights <- function(data_summary, user_context = "") {
  api_key <- Sys.getenv("OPENAI_API_KEY")
  
  if (nchar(api_key) > 0) {
    return(generate_openai_insights(data_summary, user_context, api_key))
  }
  
  # Fallback: try Ollama (local)
  return(generate_ollama_insights(data_summary, user_context))
}

#' OpenAI API integration
generate_openai_insights <- function(data_summary, user_context, api_key) {
  prompt <- sprintf(
    "You are a helpful health educator for first-generation and international female students in the U.S. 
Summarize these preventive health recommendations in plain, friendly language. Keep it under 150 words.
%s

User context: %s

Provide a brief, encouraging summary with 2-3 key takeaways.",
    data_summary,
    if (nchar(user_context) > 0) user_context else "General audience"
  )
  
  response <- tryCatch({
    result <- httr::POST(
      url = "https://api.openai.com/v1/chat/completions",
      httr::add_headers(
        "Authorization" = paste("Bearer", api_key),
        "Content-Type" = "application/json"
      ),
      body = jsonlite::toJSON(list(
        model = "gpt-4o-mini",
        messages = list(list(role = "user", content = prompt)),
        max_tokens = 300
      ), auto_unbox = TRUE),
      encode = "json"
    )
    if (httr::status_code(response) == 200) {
      content <- httr::content(response, "parsed")
      content$choices[[1]]$message$content
    } else {
      paste("AI unavailable. Status:", httr::status_code(response))
    }
  }, error = function(e) {
    paste("AI error:", e$message)
  })
  return(response)
}

#' Ollama integration (local LLM)
generate_ollama_insights <- function(data_summary, user_context) {
  prompt <- sprintf(
    "Summarize these U.S. preventive health recommendations in plain language for international students. 2-3 key takeaways. Under 150 words.\n\n%s",
    data_summary
  )
  
  response <- tryCatch({
    result <- httr::POST(
      url = "http://localhost:11434/api/generate",
      body = jsonlite::toJSON(list(
        model = "llama2",
        prompt = prompt,
        stream = FALSE
      ), auto_unbox = TRUE),
      encode = "json",
      httr::timeout(30)
    )
    if (httr::status_code(result) == 200) {
      content <- httr::content(result, "parsed")
      content$response
    } else {
      "Ollama not running. Set OPENAI_API_KEY for cloud AI, or start Ollama locally."
    }
  }, error = function(e) {
    "AI unavailable. Set OPENAI_API_KEY or run Ollama (ollama run llama2)."
  })
  return(response)
}

File: src/ai_insights.py
"""
Ollama LLM Integration - Text simplification, cultural context, summaries.
Uses Ollama Cloud (gpt-oss:120b) or local. Caches outputs for efficiency.
"""

import hashlib
import os
import requests
import re  # ✅ missing import added

# Simple in-memory cache (per session)
_llm_cache: dict[str, str] = {}


def _call_ollama(prompt: str, cache_key: str | None = None) -> str:
    """Call Ollama API with low temperature. Uses cache if key provided."""
    if cache_key and cache_key in _llm_cache:
        return _llm_cache[cache_key]

    api_key = os.getenv("OLLAMA_API_KEY")
    model = os.getenv("OLLAMA_MODEL", "gpt-oss:120b")

    if api_key:
        base_url = os.getenv("OLLAMA_HOST", "https://ollama.com")
        headers = {"Authorization": f"Bearer {api_key}"}
    else:
        base_url = os.getenv("OLLAMA_HOST", "http://localhost:11434")
        headers = {}
        model = os.getenv("OLLAMA_MODEL", "llama3.2")

    try:
        r = requests.post(
            f"{base_url.rstrip('/')}/api/chat",
            json={
                "model": model,
                "messages": [{"role": "user", "content": prompt}],
                "stream": False,
                "options": {"temperature": 0.4, "num_predict": 350},
            },
            headers=headers,
            timeout=90,
        )
        r.raise_for_status()
        data = r.json()
        content = data.get("message", {}).get("content", "").strip() or "Unable to generate."

        content = content.replace("\u2011", "").replace("\u2013", " ").replace("\u2014", " ")
        content = re.sub(r"\n\s*[-–—]\s*", "\n• ", content)
        content = re.sub(r"^\s*[-–—]\s*", "• ", content)

        if cache_key:
            _llm_cache[cache_key] = content

        return content

    except requests.exceptions.ConnectionError:
        return "Ollama unavailable. Start local Ollama or set OLLAMA_API_KEY for cloud."
    except requests.exceptions.HTTPError as e:
        if e.response and e.response.status_code == 404:
            return "Model not found. For Ollama Cloud, set OLLAMA_MODEL=gpt-oss:120b in .env"
        return f"API error: {e}"
    except Exception as e:
        return f"Error: {e}"


def simplify_text(recommendation_text: str) -> str:
    """Rewrite in plain language at 8th-grade level."""
    prompt = f"""Rewrite the following preventive health recommendation in plain language at an 8th-grade reading level. Use bullet points. Keep under 3 sentences. Remove medical jargon.

{recommendation_text}"""
    key = f"simplify_{hashlib.md5(recommendation_text[:200].encode()).hexdigest()}"
    return _call_ollama(prompt, key)


def why_this_matters(title: str, age: int) -> str:
    """Explain in 1-2 sentences why this matters for the user."""
    prompt = f"""Explain in 1-2 simple sentences why this preventive recommendation is important for a {age}-year-old international student in the U.S.

Recommendation: {title}"""

    raw = f"{title}{age}"
    key = f"why_{hashlib.md5(raw.encode()).hexdigest()}"

    return _call_ollama(prompt, key)


def cultural_context() -> str:
    """Explain preventive care differences."""
    prompt = """Briefly explain how preventive care in the U.S. may differ from other countries where people only visit doctors when sick. 2-3 sentences for international students."""
    return _call_ollama(prompt, "cultural")


def personalized_summary(recommendations_text: str, vaccination_status: list[str], age: int) -> str:
    """Short encouraging summary."""
    if not recommendations_text or not recommendations_text.strip():
        return "Based on your profile, focus on routine screenings and preventive visits. Talk to your doctor about what's right for you."

    vax = ", ".join(vaccination_status) if vaccination_status else "None specified"

    prompt = f"""Based on these preventive recommendations and vaccination status ({vax}), generate a short encouraging summary for a {age}-year-old international student.

CRITICAL: Write ONLY complete, grammatically correct sentences. Do NOT truncate or cut off mid-sentence. Each sentence must be fully finished. Keep under 100 words. Do not use bullet points or dashes.

Recommendations:
{recommendations_text}"""

    raw = f"{recommendations_text}{vax}{age}"
    key = f"summary_{hashlib.md5(raw.encode()).hexdigest()}"

    result = _call_ollama(prompt, key)

    if result and any(word in result.lower() for word in ["error", "unavailable", "not found"]):
        return f"Based on your age ({age}) and vaccination status, prioritize cervical cancer screening, blood pressure checks, and annual well-woman visits. These are key preventive steps for your health."

    return result


def myth_vs_fact() -> str:
    """Generate 3 myth vs fact pairs."""
    prompt = """Generate exactly 3 myth vs fact pairs about preventive healthcare for young women in the U.S. Format each as:

Myth: [myth]
Fact: [fact]

Keep each brief (1 sentence each)."""
    return _call_ollama(prompt, "mythfact")


# Predefined quiz items
MYTH_FACT_QUIZ = [
    ("You only need to see a doctor when you're sick.", False),
    ("Cervical cancer screening (Pap/HPV test) is recommended for women starting at age 21.", True),
    ("The flu shot can give you the flu.", False),
    ("Annual well-woman visits are recommended for preventive care.", True),
    ("HPV vaccine is only for people who are sexually active.", False),
    ("Blood pressure should be checked regularly starting at age 18.", True),
    ("You can skip vaccines if you had them as a child.", False),
    ("Mental health check-ins are part of preventive care.", True),
]
